aten/src/ATen/CPUGenerator.h

```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include <ATen/core/Generator.h>
+#include <ATen/core/PhiloxRNGEngine.h>
+
+namespace at {
+
+struct CPUGenerator;
+
+namespace detail {
+
+// Global generator state and constants
+static std::once_flag cpu_device_flag;
+static std::unique_ptr<CPUGenerator> default_gen_cpu;
+
+// Internal API
```

Comment:
Are these really internal? You call them from a lot of places... what's the normal way to get one of these generators?

Reply:
So ideally I wanted to do the following from Context.h:
```
auto getDefaultGenerator(at::CPU) -> gives me CPUGenerator
auto getDefaultGenerator(at::CUDA) -> gives me CUDAGenerator
auto getDefaultGenerator(at::device({at::kCUDA, 1}) -> gives me CUDAGenerator at device index 1
... same for CPUGenerator
```
I thought I could hack this behavior using SFINAE (similar to what I did in the PhiloxRNGEngine's operator() call) but ended giving up on it after not being able to fix template errors. So now I am creating this API inside these subclasses, such that when somebody needs a CPUGenerator, they can just call this instead of getDefaultGenerator which returns a Generator&. All this to avoid having to dynamic cast all the time in the back end, when needing specialized class methods (such as random(), random64() methods from the CPUGenerator()).

Reply:
This sounds fine, maybe they shouldn't be internal ;)